lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+  for (auto &[fanInPoint, delayAndHistory] : fanIns) {
+    auto [delay, history] = delayAndHistory;
+    auto newPath =
+        instancePathCache->prependInstance(op, fanInPoint.instancePath);
+
+    // If the fanIn is not a block argument, record it directly.
+    auto arg = dyn_cast<BlockArgument>(fanInPoint.value);
+    if (!arg) {
+      // Update the history to have correct instance path.
+      auto newHistory = debugPointFactory->getEmptyList();
+      if (ctx->doTraceDebugPoints()) {
+        newHistory =
+            mapList(debugPointFactory.get(), history, [&](DebugPoint p) {
+              p.object.instancePath = newPath;
+              return p;
+            });
+        newHistory = debugPointFactory->add(
+            DebugPoint({}, value, bitPos, delay, "output port"), newHistory);
+      }
+
+      results.emplace_back(newPath, fanInPoint.value, fanInPoint.bitPos, delay,
+                           newHistory);
+      continue;
+    }
+
+    // Otherwise, we need to look up the instance operand.
+    auto result = getOrComputeResults(op->getOperand(arg.getArgNumber()),
+                                      fanInPoint.bitPos);
+    if (failed(result))
+      return failure();
+    for (auto path : *result) {
+      auto newHistory = debugPointFactory->getEmptyList();
+      if (ctx->doTraceDebugPoints()) {
+        // Update the history to have correct instance path.
+        newHistory =
+            mapList(debugPointFactory.get(), history, [&](DebugPoint p) {
+              p.object.instancePath = newPath;
+              p.delay += path.delay;
+              return p;
+            });
+        DebugPoint debugPoint({}, value, bitPos, delay + path.delay,
+                              "output port");
+        newHistory = debugPointFactory->add(debugPoint, newHistory);
+      }
+
+      path.delay += delay;
+      path.history =
+          concatList(debugPointFactory.get(), newHistory, path.history);
+      results.push_back(path);
+    }
+  }
+
+  return success();
+}
+
+LogicalResult LocalVisitor::visit(comb::ConcatOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  // Find the exact bit pos in the concat.
+  size_t newBitPos = bitPos;
+  for (auto operand : llvm::reverse(op.getInputs())) {
+    auto size = getBitWidth(operand);
+    if (newBitPos >= size) {
+      newBitPos -= size;
+      continue;
+    }
+    return markEquivalent(op, bitPos, operand, newBitPos, results);
+  }
+
+  llvm::report_fatal_error("Should not reach here");
+  return failure();
+}
+
+LogicalResult LocalVisitor::addLogicOp(Operation *op, size_t bitPos,
+                                       SmallVectorImpl<OpenPath> &results) {
+  auto size = op->getNumOperands();
+  auto cost = std::max(1u, llvm::Log2_64_Ceil(size));
+  // Create edges each operand with cost ceil(log(size)).
+  for (auto operand : op->getOperands())
+    if (failed(addEdge(operand, bitPos, cost, results)))
+      return failure();
```

Comment:
Using the logarithm as an approximation for the ideal work depth of a logic op is a nice idea! If I remember correctly there was even some theoretical work by Torsten Hoefler showing that this holds in many cases.